### PR TITLE
ratelimit: convert to Grpc::AsyncClient.

### PR DESCRIPTION
**Pull request by htuch**

### CODE REVIEW

**Comment by htuch** [comment]
@mattklein123 This is WiP - I still need to add some more unit tests for coverage and the actual integration test that motivated this, but thought I would throw it out for comments early.

File: source/common/ratelimit/ratelimit.proto [hunk 1]
```diff
@@ -2,8 +2,6 @@ syntax = "proto3";
 
 option go_package = "ratelimit";
 
-option cc_generic_services = true;
```
Comment (mattklein123) [inline source/common/ratelimit/ratelimit.proto]: It might be out of scope for this change, but if possible can we pull this proto from envoy-api and delete this file also? There is a GH open on that.

File: source/common/ratelimit/ratelimit_impl.cc [hunk 2]
```diff
@@ -9,24 +9,28 @@
...
+          "pb.lyft.ratelimit.RateLimitService.ShouldRateLimit")),
+      async_client_(std::move(async_client)), timeout_(timeout) {}
 
 GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
 
 void GrpcClientImpl::cancel() {
-  ASSERT(callbacks_);
-  channel_->cancel();
+  ASSERT(callbacks_ != nullptr);
+  ASSERT(stream_ != nullptr);
```
Comment (mattklein123) [inline source/common/ratelimit/ratelimit_impl.cc]: nit: will crash on the next line in obvious way

File: source/common/ratelimit/ratelimit_impl.cc [hunk 3]
```diff
@@ -46,39 +50,55 @@ void GrpcClientImpl::createRequest(pb::lyft::ratelimit::RateLimitRequest& reques
...
+    callbacks_->complete(LimitStatus::Error);
+    callbacks_ = nullptr;
+    stream_ = nullptr;
+    return;
   }
 
-  callbacks_->complete(status);
-  callbacks_ = nullptr;
-}
+  LimitStatus limit_status = LimitStatus::OK;
```
Comment (mattklein123) [inline source/common/ratelimit/ratelimit_impl.cc]: Question: It seems kind of odd to do this logic in the context of onRemoteClose(). Why not onReceiveMessage() ? I guess the larger question is for unary APIs (which are very common) can we make coding against this easier vs. having to remember to close the stream, etc.?

File: source/common/ratelimit/ratelimit.proto (hunk 1 above)
Comment (htuch) [inline source/common/ratelimit/ratelimit.proto]: Yeah, I think we want to cut to the v2 API when we switch to envoy-api, and there we have a different package namespace etc., so that's best left as future work.

File: source/common/ratelimit/ratelimit_impl.cc (hunk 3 above)
Comment (htuch) [inline source/common/ratelimit/ratelimit_impl.cc]: Yep, this is a good idea, I was planning to provide a unary RPC interface that didn't require users to keep track of the minutiae of stream state before finalizing the PR.

**Comment by htuch** [comment]
@mattklein123 This now works and is ready for review. It's pretty big - if you are OK reviewing as is that's great, otherwise I can try and tease apart stuff into smaller PRs.

@msample I'm hoping this resolves your issue when using `Grpc::AsyncClient` in your custom filter.

**Comment by mattklein123** [comment]
I'm working on fixing https://github.com/lyft/envoy/issues/1254 then I can review.

File: include/envoy/grpc/async_client.h [hunk 4]
```diff
@@ -100,13 +155,12 @@ template <class RequestType, class ResponseType> class AsyncClient {
...
+   *         closeStream() is invoked by the caller to notify the client that the stream resources
+   *         may be reclaimed.
    */
-  virtual AsyncClientStream<RequestType>*
-  start(const Protobuf::MethodDescriptor& service_method,
-        AsyncClientCallbacks<ResponseType>& callbacks,
-        const Optional<std::chrono::milliseconds>& timeout) PURE;
+  virtual AsyncStream<RequestType>* start(const Protobuf::MethodDescriptor& service_method,
+                                          AsyncStreamCallbacks<ResponseType>& callbacks,
+                                          const Optional<std::chrono::milliseconds>& timeout) PURE;
```
Comment (mattklein123) [inline include/envoy/grpc/async_client.h]: Since we are now splitting streaming and unary, can we kill the timeout in the streaming case? IMO it doesn't make any sense. (I know the underlying AsyncClient stuff still has this problem, but we might as well fix it here?)

File: source/common/grpc/async_client_impl.h [hunk 5]
```diff
@@ -29,54 +31,76 @@ class AsyncClientImpl final : public AsyncClient<RequestType, ResponseType> {
...
+                                                        timeout)};
+
+    grpc_stream->initialize();
+    if (grpc_stream->hasResetStream()) {
       return nullptr;
     }
 
-    grpc_stream->set_stream(http_stream);
+    grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+    return dynamic_cast<AsyncRequestImpl<RequestType, ResponseType>*>(
```
Comment (mattklein123) [inline source/common/grpc/async_client_impl.h]: perf nit: This has cost. Typically I would just hold a local reference/pointer to the unique_ptr, then return it after you add to the list.

File: source/common/grpc/async_client_impl.h [hunk 6]
```diff
@@ -185,9 +215,10 @@ class AsyncClientStreamImpl : public AsyncClientStream<RequestType>,
 
     // This will destroy us, but only do so if we are actually in a list. This does not happen in
     // the immediate failure case.
-    if (LinkedObject<AsyncClientStreamImpl<RequestType, ResponseType>>::inserted()) {
-      LinkedObject<AsyncClientStreamImpl<RequestType, ResponseType>>::removeFromList(
-          parent_.active_streams_);
+    if (LinkedObject<AsyncStreamImpl<RequestType, ResponseType>>::inserted()) {
+      parent_.dispatcher_.deferredDelete(
```
Comment (mattklein123) [inline source/common/grpc/async_client_impl.h]: You probably already looked at this, but just double check that the stream/request does not reference client in its destructor, or there might be ordering issues depending on how client is destructed. (I haven't looked at code but just pointing this out from past experience).

File: source/common/http/async_client_impl.cc [hunk 7]
```diff
@@ -147,7 +147,7 @@ void AsyncStreamImpl::cleanup() {
   // This will destroy us, but only do so if we are actually in a list. This does not happen in
   // the immediate failure case.
   if (inserted()) {
-    removeFromList(parent_.active_streams_);
+    dispatcher().deferredDelete(removeFromList(parent_.active_streams_));
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: Same comment here about referencing client in destructor.

File: source/common/ratelimit/ratelimit_impl.h [hunk 8]
```diff
@@ -16,9 +16,16 @@
...
 
-class GrpcClientImpl : public Client, public Grpc::RpcChannelCallbacks {
+typedef Grpc::AsyncClient<pb::lyft::ratelimit::RateLimitRequest,
+                          pb::lyft::ratelimit::RateLimitResponse>
+    RateLimitAsyncClient;
+typedef std::unique_ptr<RateLimitAsyncClient> RateLimitAsyncClientPtr;
+
+typedef Grpc::AsyncRequestCallbacks<pb::lyft::ratelimit::RateLimitResponse> RateLimitAsyncCallbacks;
+
+class GrpcClientImpl : public Client, public RateLimitAsyncCallbacks {
```
Comment (mattklein123) [inline source/common/ratelimit/ratelimit_impl.h]: In looking at this code again, I realize that it's kind of dumb that we make a new "client" for every filter/request. Can you just drop a TODO in here while you are in here to optimize this at some point? (We should have thread local client per filter, and track outstanding requests).

File: source/common/grpc/async_client_impl.h [hunk 9]
```diff
@@ -49,9 +49,10 @@ class AsyncClientImpl final : public AsyncClient<RequestType, ResponseType> {
 
   AsyncStream<RequestType>* start(const Protobuf::MethodDescriptor& service_method,
                                   AsyncStreamCallbacks<ResponseType>& callbacks) override {
+    const Optional<std::chrono::milliseconds> no_timeout;
```
Comment (mattklein123) [inline source/common/grpc/async_client_impl.h]: Why was this change needed to fix ASAN? This still feels kind of busted to me. Aren't we storing a reference to the timeout inside the AsyncStreamImpl? If so should this be at client scope?

Comment (htuch) [inline source/common/grpc/async_client_impl.h]: I checked and it does get copied inside the underlying `Http::AsyncClient::Stream`, in the owned `RouteEntryImpl` https://github.com/lyft/envoy/blob/0efa18c36d6f789562690b5053c2c4b00987979e/source/common/http/async_client_impl.h#L123. 

More generally, this got me wondering about whether we should have a better convention for handling reference ownership semantics in Envoy - it's hard to tell from the type signature of a method alone whether it will store a reference for later use or copy. I think there's a fair bit of mixed use of this in Envoy.

Comment (htuch) [inline source/common/grpc/async_client_impl.h]: The `Grpc::AsyncStreamImpl` was not owning the timeout in period between object construction and passing to the `Http::AsyncClient::Stream` in `initialize()`, which is what triggered the ASAN fail.

Comment (mattklein123) [inline source/common/grpc/async_client_impl.h]: I see, it gets used in the context of initialize(). Sure open to suggestions on naming. Same problem applies to pointers or references. Not sure if there is a good naming scheme.

Comment (htuch) [inline source/common/grpc/async_client_impl.h]: A first thought is `FooBar&` for the case when a reference is not retained past the unwinding of the call stack, `FooBarRetainedRef` for when it is. But, that's super verbose. @dnoe for thoughts as well.
